### ISSUE TITLE
Generate DSL documentation from metamodel (grammar)

### ISSUE THREAD

**Issue by alensuljkanovic** [textX/textX#148]
Once DSL is created, documentation still needs to be written manually. 

If somehow we could document elements inside grammar, documentation will be part of metamodel (?). This would allow us to generate documentation in different formats (markdown, xml, html, ...). Plus, I believe that it would be easier to maintain the documentation, since it's in the same file as a language definition.

An interesting paper on this topic can be found [here](https://www.sciencedirect.com/science/article/pii/S1477842417301811).

**Comment by goto40** [comment]
Do you mean something like Javadoc for grammars...?  The referenced paper seems to describe another approach.

**Comment by igordejanovic** [comment]
Sounds interesting. I think it is worth researching the possibilities.

I guess something like javadoc approach would be relatively easy to implement. We already have [textX grammar defined in textX](https://github.com/textX-tools/textX-languageserver/blob/master/src/metamodel/textx.tx). :)

**Comment by alensuljkanovic** [comment]
@goto40 actually, yes. Support for comment blocks is already there, so they could be used as docstrings.

**Comment by Aluriak** [comment]
Hello !

I want to tackle this problem, since i needed it few days ago, and will probably need it in the future.
I made a first proof of concept [here](https://github.com/aluriak/textx-dsldoc), showing that something is definitively possible (although it seems that most code is a hand-crafted way to explore the structures… is there any standard visitor implementation ?).

I would like to open the discussion about (1) integration into textx and (2) methods to explore the grammar/parser i do not know about.

End goal is obviously to get a good doc generator.

**Comment by goto40** [comment]
Great!

I need some time to look at it. 

Quick idea: Showing also the type of the rule (e.g. abstract) would greatly improve the overall compréhension of a grammar. You probably already integrated this idea.

**Comment by Aluriak** [comment]
Take your time ; maybe just take a quick look to the output html, see what information is missing, that would make for a first feedback.

How can i detect the abstract type ?

**Comment by igordejanovic** [comment]
@Aluriak Looks really good so far! It would be really nice to integrate it as a part of `textx` command. I guess the best way for this and all further sub-commands would be to have an extensibility mechanism in place (something very similar to what is implemented in [textx-tools](https://github.com/textX/textx-tools/blob/master/setup.py#L59)). In that way all additional sub-command could be an optional dependencies of textX and if installed via pip would be visible and accessible.

@goto40 
> Quick idea: Showing also the type of the rule (e.g. abstract) would greatly improve the overall compréhension of a grammar.

Good idea. Maybe to state both abstract and match rules.

@Aluriak 
> How can i detect the abstract type ?

Each meta-class have a `_tx_type` attribute which value is either `abstract`, `common` or `match`.
You can use meta-model as a dict keyed by meta-class name.
```
metamodel['MyRule']._tx_type
```
To take into account and investigate grammar importing see `metamodel.namespaces`.

**Comment by igordejanovic** [comment]
@Aluriak BTW, it is still in early stage but I feel that the `core` part of [textX-LS](https://github.com/textX/textX-LS) should have a lot in common to the analysis geared towards docs generation.

**Comment by danixeee** [comment]
I agree with @igordejanovic, maybe we can add the command to `textX-LS-core` package. I have already thought about `hover` capability of language server. It would be nice to show grammar rule docstring on mouse hover, like [here](https://raw.githubusercontent.com/palantir/python-language-server/develop/resources/hover.gif).

I am currently working on support for `textxfile` language and I have pretty much copied [CLI stuff](https://github.com/textX/textX-LS/blob/danixeee/textxfile-lang-support/textX-LS/core/setup.py#L42) from `textx-tools` project.

Do we want to extend `textx` CLI with `textX-LS-core` subcommands or we should keep them separate?

**Comment by igordejanovic** [comment]
> Do we want to extend textx CLI with textX-LS-core subcommands or we should keep them separate?

I think that it is easier for a user to remember just one command `textx` and then query it for further sub-commands. It is prevailing way of work with contemporary complex tools (think `git` for example). LS sub-commands could be grouped under e.g. `ls` sub-command (`textx ls some_command...`).

**Comment by igordejanovic** [comment]
Moving discussion regarding textx CLI extension to #161 .

**Comment by Aluriak** [comment]
> @Aluriak BTW, it is still in early stage but I feel that the `core` part of [textX-LS](https://github.com/textX/textX-LS) should have a lot in common to the analysis geared towards docs generation.

Cool. Let's do that well. When do you think you can work on that ? I am obviously a future user of that core, so i'm eager to get info. Until then, i will rely on my own visitors (less classes, more functions, i'm not an oop guy), and maybe that will inspire you.

**Comment by igordejanovic** [comment]
Great! @danixeee started textX-LS recently and is actively working on it so I guess the two of you could see if and how you could integrate your efforts. I'll see to provide necessary pieces to integrate CLI commands into textX (issue #161).

**Comment by Aluriak** [comment]
And… It's done !

Look at [the repository](https://github.com/aluriak/textx-dsldoc), or `pip install textx-dsldoc` (when textx==1.9.0 will be out).
